### PR TITLE
Fcc 124 design schedule for importers

### PR DESCRIPTION
**Pull request by kpennels**
**New_files.py**

importers/scheduler_new.py - Schedule main_task to run once a day, which in turn schedules each of the importers that are present in the api table according to their refresh time.

### CODE REVIEW

File: Analytics/importers/scheduler_new.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
+import os, sys
+from datetime import datetime, timedelta
+import time
+import importlib
+import logging
+sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))
```
Comment (j0rdanl0urens) [inline Analytics/importers/scheduler_new.py]: This should be at the end of the imports by logging

File: Analytics/importers/scheduler_new.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+
+class Scheduler(object):
+    """ Schedule the execution importer tasks """
+
+    def __init__(self):
+        """ Create SQLAlchemy session and launch APScheduler """
+
+        self.engine = sqlalchemy.create_engine(settings.DB_URI)
+        _session = scoped_session(sessionmaker(bind=self.engine))
```
Comment (j0rdanl0urens) [inline Analytics/importers/scheduler_new.py]: Is this variable set just to be assigned in the next line and nothing else?

Comment (j0rdanl0urens) [reply Analytics/importers/scheduler_new.py]: If so please remove it

File: Analytics/importers/scheduler_new.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    def get_apis(self) -> list:
+        """
+        Retrieve API details stored in the database
+        :return: a list containing the details of APIs
+        """
+
+        _api = self.session.execute('select * from api')
+        apis = []
+        for api in _api.fetchall():
+            a = API(name=api[1], url=api[2], api_key=api[3],
```
Comment (j0rdanl0urens) [inline Analytics/importers/scheduler_new.py]: single letter variables are bad variables

File: Analytics/importers/scheduler_new.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    def fetch_data(self, class_name: str, api_name: str):
+        """
+        Import sensor data from the specified API endpoint
+        :param class_name: name of the class that implements the
+        corresponding importer
+        :param api_name: identifying name stored in the database for the
+        respective importer
+        """
+
+        from app import create_app
```
Comment (j0rdanl0urens) [inline Analytics/importers/scheduler_new.py]: Why is this not in the imports at the top of the file?

File: Analytics/importers/scheduler_new.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        from db import db
+        a = create_app()
```
Comment (j0rdanl0urens) [inline Analytics/importers/scheduler_new.py]: Rename

File: Analytics/importers/scheduler_new.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+        self.api_class = api_class
+        self.refresh_time = refresh_time
+        self.token_expiry = token_expiry
+        self.timestamp = timestamp
+
+
+if __name__ == '__main__':
+    scheduler = Scheduler()
+    scheduler.run()
+
```
Comment (j0rdanl0urens) [inline Analytics/importers/scheduler_new.py]: Is this 2 new lines?

File: Analytics/importers/scheduler_new.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    def __init__(self, name: str, url: str, api_key: str, api_class: str,
+                 refresh_time: str, token_expiry: datetime, timestamp: datetime):
+        """
+        Store fields of api table in a class
+        :param name: name of the importer
+        :param url: API endpoint for accessing data to import
+        :param api_key: key which authorises access to the API endpoint
+        :param api_class: name of the class which implements the importer
+        :param refresh_time: number of seconds to wait until running the
+        importer again
```
Comment (j0rdanl0urens) [inline Analytics/importers/scheduler_new.py]: Please indent this

File: Analytics/importers/scheduler_new.py [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+    def main_task(self):
+        """
+        Schedule the executions for each of the importers at their specified
+        refresh time
+        """
+
+        apis = self.get_apis()
+        for api in apis:
+            self.scheduler.add_job(self.fetch_data, 'interval',
```
Comment (j0rdanl0urens) [inline Analytics/importers/scheduler_new.py]: Please re-align this properly

File: Analytics/importers/scheduler_new.py (hunk 4 above)
Comment (kpennels) [reply Analytics/importers/scheduler_new.py]: The fetch_data function was implemented by an FCC developer so I can't exactly answer this question. But when I do try to place the imports at the top of the file, the import process fails

**Pull request merged**